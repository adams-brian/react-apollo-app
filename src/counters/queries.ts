import { gql } from 'apollo-boost';

export const CountersQuery = gql`
  query {
    counters
  }
`;

export interface ICountersQueryResponse {
  counters: number[]
}

export const SaveCountersMutation = gql`
  mutation saveCounters( $counters: [Int]! ) {
    saveCounters( counters: $counters )
  }
`;

export interface ISaveCountersVariables {
  counters: number[]
}
