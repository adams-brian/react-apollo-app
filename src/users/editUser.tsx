import { History } from 'history';
import * as React from 'react';
import { graphql } from 'react-apollo';
import { RouteComponentProps } from 'react-router';
import { Link, withRouter } from 'react-router-dom';
import { branch, mapProps, renderComponent } from 'recompose';

import Loading from '../common/loading';
import { CREATE_USER_TEMP_ID, CreateUserMutation, ICreateUserResponse,
  ICreateUserVariables, IUpdateUserResponse, IUser,
  IUserQueryResponse, IUsersQueryResponse,
  IUserVariables, TCreateUserFunc, TUpdateUserFunc,
  UpdateUserMutation, UserQuery, UsersQuery } from './queries';

interface IProps {
  createUser: TCreateUserFunc;
  history: History;
  updateUser: TUpdateUserFunc;
  user: IUser;
}

export class EditUser extends React.Component<IProps, IUser> {
  constructor(props: IProps) {
    super(props);
    this.state = {...props.user};
  }

  public render() {
    return (
      <div className="edit-user">
        <h1>{this.state.id && this.state.id.length > 0 ? 'Edit' : 'Create'} User</h1>
        <form>
          <div className="form-group">
            <label htmlFor="firstname">Firstname</label>
            <input
              placeholder="Firstname"
              type="text"
              className="form-control"
              name="firstname"
              id="firstname"
              value={this.state.firstname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div className="form-group">
            <label htmlFor="lastname">Lastname</label>
            <input
              placeholder="Lastname"
              type="text"
              className="form-control"
              name="lastname"
              id="lastname"
              value={this.state.lastname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div>
            <Link to="/users">
              <button type="button" className="btn btn-secondary">Cancel</button>
            </Link>
            {' '}
            <button 
              className="btn btn-primary" 
              type="submit" 
              onClick={this.submit} 
              disabled={this.state.firstname.length === 0 || this.state.lastname.length === 0}
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }

  private inputChanged = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({
      [e.currentTarget.name]: e.currentTarget.value
    } as Pick<IUser, keyof IUser>);
  }

  private submit = (e: React.FormEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.state.id && this.state.id.length > 0) {
      this.props.updateUser({
        optimisticResponse: {
          updateUser: { ...this.state, __typename: 'User' }
        },
        variables: { ...this.state }
      });
    }
    else {
      this.props.createUser({
        optimisticResponse: {
          createUser: { ...this.state, id: CREATE_USER_TEMP_ID, __typename: 'User' }
        },
        update: (store, data) => {
          if (data.data) {
            const user = data.data.createUser;
            const users = store.readQuery<IUsersQueryResponse>({ query: UsersQuery });
            const list = users && users.users ? users.users : [];
            store.writeQuery({
              data: { users: [...list, {...user}] },
              query: UsersQuery
            });
          }
        },
        variables: { firstname: this.state.firstname, lastname: this.state.lastname } });
    }
    this.props.history.push('/users');
  }
}

export const generateComponent = () =>
  withRouter(
    mapProps(
      (props: RouteComponentProps<{id: string}>) => ({ history: props.history, id: props.match.params.id })
    )(
      graphql<{ history: History, id: string }, IUserQueryResponse, IUserVariables, { history: History, id: string, loading?: boolean, user?: IUser }>(UserQuery, {
        options: (props) => ({
          variables: {
            id: props.id
          }
        }),
        props: (props) => ({
          history: props.ownProps.history,
          id: props.ownProps.id,
          loading: props.data === undefined || props.data.loading === true,
          user: props.data !== undefined && props.data.user !== undefined ? props.data.user : { id: '', firstname: '', lastname: '' }
        }),
        skip: (props) => !(props.id && props.id.length > 0)
      })(
        branch((props: { history: History, id: string, loading?: boolean, user?: IUser }) => props.loading !== undefined && props.loading === true,
          renderComponent(Loading)
        )(
          mapProps(
            (props: { history: History, id: string, loading?: boolean, user?: IUser }) => ({
              history: props.history,
              user: props.user !== undefined ? props.user : { id: '', firstname: '', lastname: '' }
            })
          )(
            graphql<{ history: History, user: IUser }, ICreateUserResponse, ICreateUserVariables, { createUser: TCreateUserFunc, history: History, user: IUser }> (CreateUserMutation, {
              props: (props) => ({
                createUser: props.mutate!,
                history: props.ownProps.history,
                user: props.ownProps.user
              })
            })(
              graphql<{ createUser: TCreateUserFunc, history: History, user: IUser }, IUpdateUserResponse, IUser, IProps> (UpdateUserMutation, {
                props: (props) => ({
                  createUser: props.ownProps.createUser,
                  history: props.ownProps.history,
                  updateUser: props.mutate!,
                  user: props.ownProps.user
                })
              })(
                EditUser
              )
            )
          )
        )
      )
    )
  )

export default generateComponent();
