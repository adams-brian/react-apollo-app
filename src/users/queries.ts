import { gql } from 'apollo-boost';
import { MutationFunc } from 'react-apollo';

export interface IUser {
  id: string;
  firstname: string;
  lastname: string;
}

interface IAddTypename {
  __typename: string;
}

export const UsersQuery = gql`
  query {
    users {
      id
      firstname
      lastname
    }
  }
`;
export interface IUsersQueryResponse {
  users: IUser[];
}

export const UserQuery = gql`
  query user ( $id: ID! ) {
    user( id: $id ) {
      id
      firstname
      lastname
    }
  }
`;
export interface IUserVariables {
  id: string;
}
export interface IUserQueryResponse {
  user: IUser;
}

export const CREATE_USER_TEMP_ID = "[TEMPORARY ID]";
export const CreateUserMutation = gql`
  mutation createUser( $firstname: String!, $lastname: String! ) {
    createUser( firstname: $firstname, lastname: $lastname ) {
      id
      firstname
      lastname
    }
  }
`;
export interface ICreateUserVariables {
  firstname: string;
  lastname: string;
}
export interface ICreateUserResponse {
  createUser: IUser & IAddTypename;
}
export type TCreateUserFunc = MutationFunc<ICreateUserResponse, ICreateUserVariables>;

export const UpdateUserMutation = gql`
  mutation updateUser( $id: ID!, $firstname: String!, $lastname: String! ) {
    updateUser(id: $id, firstname: $firstname, lastname: $lastname) {
      id
      firstname
      lastname
    }
  }
`;
export interface IUpdateUserResponse {
  updateUser: IUser & IAddTypename;
}
export type TUpdateUserFunc = MutationFunc<IUpdateUserResponse, IUser>;

export const DeleteUserMutation = gql`
  mutation deleteUser( $id: ID! ) {
    deleteUser( id: $id ) {
      id
    }
  }
`;
export interface IDeleteUserVariables {
  id: string;
}
export interface IDeleteUserResponse {
  deleteUser: {
    id: string
  } & IAddTypename;
}
export type TDeleteUserFunc = MutationFunc<IDeleteUserResponse, IDeleteUserVariables>;
