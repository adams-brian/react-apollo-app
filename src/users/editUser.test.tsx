import { shallow } from 'enzyme';
import { createMemoryHistory, MemoryHistory } from 'history'
import * as React from 'react';
import * as reactApollo from 'react-apollo';
import * as reactRouterDom from 'react-router-dom';
import * as recompose from 'recompose';
import * as sinon from 'sinon';

import Loading from '../common/loading';
import { EditUser, generateComponent } from './editUser';
import { CreateUserMutation, UpdateUserMutation, UserQuery, UsersQuery } from './queries';

const users = [
  {
    firstname: 'abc',
    id: '123',
    lastname: 'def'
  },
  {
    firstname: 'hij',
    id: '456',
    lastname: 'klm'
  }
];

describe("EditUser", () => {

  describe('component', () => {

    let element: JSX.Element;
    let history: MemoryHistory;
    let push: sinon.SinonSpy;
    let createUser: sinon.SinonSpy;
    let updateUser: sinon.SinonSpy;

    beforeEach(() => {
      history = createMemoryHistory();
      push = sinon.spy(history, 'push');
      createUser = sinon.spy();
      updateUser = sinon.spy();
    });

    afterEach(() => {
      push.restore();
    });

    describe("create user mode", () => {

      beforeEach(() => {
        element = <EditUser
          createUser={createUser}
          history={history}
          updateUser={updateUser}
          user={{
            firstname: '',
            id: '',
            lastname: ''
          }}
        />;
      });

      it('renders as expected', () => {
        const component = shallow(element);
        expect(component).toMatchSnapshot();
      });

      it('calls createUser on submit', () => {
        const component = shallow(element);
        expect(createUser.called).toBe(false);
        expect(push.called).toBe(false);
        const preventDefault = sinon.spy();
        const user = {
          firstname: 'new',
          id: 'newid',
          lastname: 'user'
        };
        component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: user.firstname } });
        component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: user.lastname } });
        component.find('button[type="submit"]').first().simulate('click', { preventDefault });
        expect(preventDefault.called).toBe(true);
        expect(createUser.calledOnce).toBe(true);
        expect(createUser.firstCall.args.length).toBe(1);
        const config = createUser.firstCall.args[0];
        expect(config).toMatchSnapshot();
        let readQuery = sinon.fake.returns({ users });
        const writeQuery = sinon.spy();
        config.update({ readQuery, writeQuery }, { data: { createUser: {...user, __typename: 'User'}}});
        expect(readQuery.calledOnceWithExactly({ query: UsersQuery })).toBe(true);
        expect(writeQuery.calledOnceWithExactly({
          data: { users: [...users, {...user, __typename: 'User'}]},
          query: UsersQuery
        })).toBe(true);
        readQuery = sinon.fake.returns({ });
        writeQuery.resetHistory();
        config.update({ readQuery, writeQuery }, {});
        config.update({ readQuery, writeQuery }, { data: { createUser: {...user, __typename: 'User'}}});
        expect(readQuery.calledOnceWithExactly({ query: UsersQuery })).toBe(true);
        expect(writeQuery.calledOnceWithExactly({
          data: { users: [{...user, __typename: 'User'}]},
          query: UsersQuery
        })).toBe(true);
        expect(push.calledOnceWithExactly('/users')).toBe(true);
      });

    });

    describe("edit user mode", () => {

      beforeEach(() => {
        element = <EditUser
          createUser={createUser}
          history={history}
          updateUser={updateUser}
          user={{
            firstname: 'abc',
            id: '123',
            lastname: 'def'
          }}
        />;
      });

      it('renders as expected', () => {
        const component = shallow(element);
        expect(component).toMatchSnapshot();
      });

      it('calls updateUser', () => {
        const component = shallow(element);
        expect(updateUser.called).toBe(false);
        expect(push.called).toBe(false);
        const preventDefault = sinon.spy();
        const user = {
          firstname: 'updated',
          id: '123',
          lastname: 'user'
        };
        component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: user.firstname } });
        component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: user.lastname } });
        component.find('button[type="submit"]').first().simulate('click', { preventDefault });
        expect(preventDefault.called).toBe(true);
        expect(updateUser.calledOnceWithExactly({
          optimisticResponse: {
            updateUser: {...user, __typename: 'User'}
          },
          variables: user
        })).toBe(true);
        expect(push.calledOnceWithExactly('/users')).toBe(true);
      });

    });

  });

  describe('connection', () => {

    let withRouter: sinon.SinonStub;
    let graphql: sinon.SinonStub;
    let branch: sinon.SinonStub;
    let mapProps: sinon.SinonStub;
    let renderComponent: sinon.SinonStub;

    const replaceGqlNames = (arg: any) => {
      if (arg === CreateUserMutation) {
        return 'CreateUserMutation';
      }
      if (arg === UpdateUserMutation) {
        return 'UpdateUserMutation';
      }
      if (arg === UserQuery) {
        return 'UserQuery';
      }
      return arg;
    }
    const tracker = (name: string) =>
      (...setupArgs: any[]) =>
        (...callArgs: any[]) => ({
          callArgs: callArgs.map(a => replaceGqlNames(a)),
          name,
          setupArgs: setupArgs.map(a => replaceGqlNames(a)),
        })

    beforeEach(() => {
      withRouter = sinon.stub(reactRouterDom, 'withRouter').callsFake( tracker('withRouter')() );
      graphql = sinon.stub(reactApollo, 'graphql').callsFake( tracker('graphql') );
      branch = sinon.stub(recompose, 'branch').callsFake( tracker('branch') );
      mapProps = sinon.stub(recompose, 'mapProps').callsFake( tracker('mapProps') );
      renderComponent = sinon.stub(recompose, 'renderComponent').callsFake(
        (...args) =>
          args.length === 1 && args[0] === Loading? 'renderComponent(Loading)' : 'renderComponent(UNKNOWN)'
      );
    });

    afterEach(() => {
      withRouter.restore();
      graphql.restore();
      branch.restore();
      mapProps.restore();
      renderComponent.restore();
    });

    it('connects as expected', () => {
      const composition = generateComponent() as any;
      expect(composition).toMatchSnapshot();

      let current = composition;

      expect(current.name).toBe('withRouter');
      expect(current.setupArgs).toEqual([]);
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('mapProps');
      expect(current.setupArgs.length).toBe(1);
      let mapper = current.setupArgs[0];
      expect(mapper({
        history: 'the history',
        match: { params: { id: 'the id' } }
      })).toEqual({
        history: 'the history',
        id: 'the id'
      });
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('graphql');
      expect(current.setupArgs.length).toBe(2);
      expect(current.setupArgs[0]).toBe('UserQuery');
      let opts = current.setupArgs[1];
      expect(opts.options({ id: 'the id' })).toEqual({
        variables: { id: 'the id' }
      });
      expect(opts.props({
        ownProps: {
          history: 'the history',
          id: 'the id'
        }
      })).toEqual({
        history: 'the history',
        id: 'the id',
        loading: true,
        user: { id: '', firstname: '', lastname: '' }
      });
      expect(opts.props({
        data: { },
        ownProps: {
          history: 'the history',
          id: 'the id'
        }
      })).toEqual({
        history: 'the history',
        id: 'the id',
        loading: false,
        user: { id: '', firstname: '', lastname: '' }
      });
      expect(opts.props({
        data: {
          user: 'the user'
        },
        ownProps: {
          history: 'the history',
          id: 'the id'
        }
      })).toEqual({
        history: 'the history',
        id: 'the id',
        loading: false,
        user: 'the user'
      });
      expect(opts.props({
        data: {
          loading: true,
          user: 'the user'
        },
        ownProps: {
          history: 'the history',
          id: 'the id'
        }
      })).toEqual({
        history: 'the history',
        id: 'the id',
        loading: true,
        user: 'the user'
      });
      expect(opts.props({
        data: {
          loading: false,
          user: 'the user'
        },
        ownProps: {
          history: 'the history',
          id: 'the id'
        }
      })).toEqual({
        history: 'the history',
        id: 'the id',
        loading: false,
        user: 'the user'
      });
      expect(opts.skip({})).toBe(true);
      expect(opts.skip({ id: null })).toBe(true);
      expect(opts.skip({ id: '' })).toBe(true);
      expect(opts.skip({ id: 'the id' })).toBe(false);
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('branch');
      expect(current.setupArgs.length).toBe(2);
      const test = current.setupArgs[0];
      expect(test({ })).toBe(false);
      expect(test({ otherStuff: 'anything' })).toBe(false);
      expect(test({ otherStuff: 'anything', loading: false })).toBe(false);
      expect(test({ otherStuff: 'anything', loading: true })).toBe(true);
      expect(current.setupArgs[1]).toBe('renderComponent(Loading)');
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('mapProps');
      expect(current.setupArgs.length).toBe(1);
      mapper = current.setupArgs[0];
      expect(mapper({
        history: 'the history',
        id: 'the id',
      })).toEqual({
        history: 'the history',
        user: { id: '', firstname: '', lastname: '' }
      });
      expect(mapper({
        history: 'the history',
        id: 'the id',
        loading: true,
        user: 'the user'
      })).toEqual({
        history: 'the history',
        user: 'the user'
      });
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('graphql');
      expect(current.setupArgs.length).toBe(2);
      expect(current.setupArgs[0]).toBe('CreateUserMutation');
      opts = current.setupArgs[1];
      expect(opts.props({
        mutate: 'createUser',
        ownProps: {
          history: 'the history',
          user: 'the user'
        }
      })).toEqual({
        createUser: 'createUser',
        history: 'the history',
        user: 'the user'
      });
      expect(current.callArgs.length).toBe(1);

      current = current.callArgs[0];
      expect(current.name).toBe('graphql');
      expect(current.setupArgs.length).toBe(2);
      expect(current.setupArgs[0]).toBe('UpdateUserMutation');
      opts = current.setupArgs[1];
      expect(opts.props({
        mutate: 'updateUser',
        ownProps: {
          createUser: 'createUser',
          history: 'the history',
          user: 'the user'
        }
      })).toEqual({
        createUser: 'createUser',
        history: 'the history',
        updateUser: 'updateUser',
        user: 'the user'
      });
      expect(current.callArgs.length).toBe(1);
      expect(current.callArgs[0]).toBe(EditUser);

    });

  });

});
