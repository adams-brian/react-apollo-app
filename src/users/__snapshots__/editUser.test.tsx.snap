// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser component create user mode calls createUser on submit 1`] = `
Object {
  "optimisticResponse": Object {
    "createUser": Object {
      "__typename": "User",
      "firstname": "new",
      "id": "[TEMPORARY ID]",
      "lastname": "user",
    },
  },
  "update": [Function],
  "variables": Object {
    "firstname": "new",
    "lastname": "user",
  },
}
`;

exports[`EditUser component create user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Create
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`EditUser component edit user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Edit
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value="abc"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value="def"
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`EditUser connection connects as expected 1`] = `
Object {
  "callArgs": Array [
    Object {
      "callArgs": Array [
        Object {
          "callArgs": Array [
            Object {
              "callArgs": Array [
                Object {
                  "callArgs": Array [
                    Object {
                      "callArgs": Array [
                        Object {
                          "callArgs": Array [
                            [Function],
                          ],
                          "name": "graphql",
                          "setupArgs": Array [
                            "UpdateUserMutation",
                            Object {
                              "props": [Function],
                            },
                          ],
                        },
                      ],
                      "name": "graphql",
                      "setupArgs": Array [
                        "CreateUserMutation",
                        Object {
                          "props": [Function],
                        },
                      ],
                    },
                  ],
                  "name": "mapProps",
                  "setupArgs": Array [
                    [Function],
                  ],
                },
              ],
              "name": "branch",
              "setupArgs": Array [
                [Function],
                "renderComponent(Loading)",
              ],
            },
          ],
          "name": "graphql",
          "setupArgs": Array [
            "UserQuery",
            Object {
              "options": [Function],
              "props": [Function],
              "skip": [Function],
            },
          ],
        },
      ],
      "name": "mapProps",
      "setupArgs": Array [
        [Function],
      ],
    },
  ],
  "name": "withRouter",
  "setupArgs": Array [],
}
`;
